### PR TITLE
ctsm5.3.059: Various cleanup efforts from the cesm3_0_beta04 tags for testing and usability

### LINKED ISSUE #2762: Don't allow SP, fates, or nofire tests to turn fire_emis on (and make sure our tests don't do this either)
When -fire_emis is added to CLM_BLDNML_OPTS it should fail for configurations where fire-emissions won't be passed back. Right now that's SP mode, when nofire is turned on for BGC, or for FATES.

The FORTRAN code and build-namelist should disallow these options.

And we should make sure our testing don't turn on fire-emissions for these configurations.

Related issue: #2759

Definition of done:
- [x] Get the PF unit tests started
- [x] Update submodules so that full ESMF is available
- [x] Finish the unit tests with full ESMF
- [x] Make the needed changes in FORTRAN source and continue to update the unit tests

### LINKED ISSUE #2768: Add ne3np4 to list of surface dataset resolutions
For support of SCAM we need to support ne3np4 in addition to ne3np4.pg3 as pointed out by @jtruesdal.

Definition of done:
- [x] Create the needed files from a branch tag of ctsm5.3.0 (do rest on b4b-dev)
- [x] Add build of ne3np4 for same configurations as ne3np4.pg3 (1850, 2000, Hist, SSP2-4.5) to mksurfdata_esmf Makefile
- [x] Add the files into namelist_defaults
- [x] Add a test for it to the build-namelist tester
- [x] Add a SCAM test using it to aux_clm (and ctsm_sci?)

### LINKED ISSUE #2780: New CN matrix fails with single point sites with the new ctsm5.3 datasets.
Since ctsm5.2.dev175 to ctsm5.3.0 we've been running tests with MIMICS with above ground CN matrtix that have been passing. The test is SMS_D.1x1_brazil.I1850Clm60BgcCrop.derecho_gnu.clm-mimics_matrixcn. This has the soil CN matrix off (because MIMICS is non-linear), but above ground CN matrix on (use_soil_matrixcn = .false. use_matrixcn = .true.).

There are two reasons for doing this test:

1. Hopefully get MIMICS to spinup faster with above ground matrix on
2. More extensive testing of Matrix for an edge case where it might fail easier

The hope for "1" was especially there as we weren't finding methods to speed up the spinup of MIMICS. The test did pass for 30 tags, and just started failing in ctsm5.3.0 with the following type of error in the log files: 

lnd.log:
```
 hist_htapes_wrapup : Closing local history file ./SMS_D.1x1_brazil.I1850Clm60BgcCrop.derecho_gnu.clm-mimics_matrixcn.20240923_125029_ialh14.clm2.h1.0001-01-01-28800.nc at nstep =           16

(shr_strdata_readstrm) reading file ub: /glade/campaign/cesm/cesmdata/inputdata/atm/datm7/NASA_LIS/clmforc.Li_2016_climo1995-2013.360x720.lnfm_Total_c160825.nc       7
 ERROR: ERROR in /glade/work/erik/ctsm_worktrees/answer_changes/src/utils/SparseMatrixMultiplyMod.F90 at line 1246
```

cesm.log:
```
dec0996.hsn.de.hpc.ucar.edu 0:  ERROR: ERROR in /glade/work/erik/ctsm_worktrees/answer_changes/src/utils/SparseMatrixMultiplyMod.F90 at line 1246
dec0996.hsn.de.hpc.ucar.edu 0: #0  0x12c3b50 in __shr_abort_mod_MOD_shr_abort_backtrace
dec0996.hsn.de.hpc.ucar.edu 0: 	at /glade/work/erik/ctsm_worktrees/answer_changes/share/src/shr_abort_mod.F90:104
dec0996.hsn.de.hpc.ucar.edu 0: #1  0x12c3c13 in __shr_abort_mod_MOD_shr_abort_abort
dec0996.hsn.de.hpc.ucar.edu 0: 	at /glade/work/erik/ctsm_worktrees/answer_changes/share/src/shr_abort_mod.F90:61
dec0996.hsn.de.hpc.ucar.edu 0: #2  0x131f9c8 in __shr_assert_mod_MOD_shr_assert
dec0996.hsn.de.hpc.ucar.edu 0: 	at /glade/work/erik/ctsm_worktrees/answer_changes/share/src/shr_assert_mod.F90.in:95
dec0996.hsn.de.hpc.ucar.edu 0: #3  0xe38814 in __sparsematrixmultiplymod_MOD_spmp_abc
dec0996.hsn.de.hpc.ucar.edu 0: 	at /glade/work/erik/ctsm_worktrees/answer_changes/src/utils/SparseMatrixMultiplyMod.F90:1246
dec0996.hsn.de.hpc.ucar.edu 0: #4  0x8e97db in __cnvegmatrixmod_MOD_cnvegmatrix
dec0996.hsn.de.hpc.ucar.edu 0: 	at /glade/work/erik/ctsm_worktrees/answer_changes/src/biogeochem/CNVegMatrixMod.F90:1509
dec0996.hsn.de.hpc.ucar.edu 0: #5  0x10466ef in __cndrivermod_MOD_cndriverleaching
dec0996.hsn.de.hpc.ucar.edu 0: 	at /glade/work/erik/ctsm_worktrees/answer_changes/src/biogeochem/CNDriverMod.F90:1098
dec0996.hsn.de.hpc.ucar.edu 0: #6  0x92a6b2 in __cnvegetationfacade_MOD_ecosystemdynamicspostdrainage
dec0996.hsn.de.hpc.ucar.edu 0: 	at /glade/work/erik/ctsm_worktrees/answer_changes/src/biogeochem/CNVegetationFacade.F90:1125
dec0996.hsn.de.hpc.ucar.edu 0: #7  0x5d7ed6 in __clm_driver_MOD_clm_drv
dec0996.hsn.de.hpc.ucar.edu 0: 	at /glade/work/erik/ctsm_worktrees/answer_changes/src/main/clm_driver.F90:1119
```

The line it fails on from above is the SHR_ASSERT_FL in this section of code in SparseMatrixMultiplyMod.F90:

``` fortran
    if(present(num_actunit_C))then
       if(num_actunit_C < 0)then
          write(iulog,*) "error: num_actunit_C cannot be less than 0"
          call endrun( subname//" ERROR: bad value for num_actunit_C" )
          return
       end if
       if(.not. present(filter_actunit_C))then
          write(iulog,*) "error: num_actunit_C is presented but filter_actunit_C is missing"
          call endrun( subname//" ERROR: missing required optional arguments" )
          return
       end if
       SHR_ASSERT_FL((size(filter_actunit_C) > num_actunit_C), sourcefile, __LINE__)
    end if
```

The call in CNVegMatrixMod.F90 is here:

``` fortran
         if(num_actfirep .eq. 0 .and. nthreads < 2)then
            call AKallvegc%SPMP_AB(num_soilp,filter_soilp,AKphvegc,AKgmvegc,list_ready_phgmc,list_A=list_phc_phgm,list_B=list_gmc_phgm,&
                 NE_AB=NE_AKallvegc,RI_AB=RI_AKallvegc,CI_AB=CI_AKallvegc)
         else
            call AKallvegc%SPMP_ABC(num_soilp,filter_soilp,AKphvegc,AKgmvegc,AKfivegc,list_ready_phgmfic,list_A=list_phc_phgmfi,&
                 list_B=list_gmc_phgmfi,list_C=list_fic_phgmfi,NE_ABC=NE_AKallvegc,RI_ABC=RI_AKallvegc,CI_ABC=CI_AKallvegc,&
                 use_actunit_list_C=.True.,num_actunit_C=num_actfirep,filter_actunit_C=filter_actfirep)
         end if
```

Definition of done:
- [x] FAIL: Test if works for cold start
- [x] NO: Assess if should add a short f10 test and make sure it works
- [x] Change accordingly to what is found out from above

### LINKED ISSUE #2791: Add f19 16pft 1850 and Hist to list of surface dataset resolutions (for the PPE work)
To be consistent with the PPE work we need f19 datasets for 1850 and Historical transient for 16pfts. These datasets will first be used for the f19 spinup needed.

@olyson @djk2120 @linniahawkins 

Definition of done:
- [x] Create the datasets
- [x] rimport them
- [x] Add them to namelist_defaults
- [x] Add them for automatic creation in the mksurfdata_esmf Makefile

### LINKED ISSUE #2868: Update "custom crop calendars" instructions
[Here](https://escomp.github.io/ctsm-docs/versions/master/html/users_guide/running-special-cases/running-with-custom-crop-calendars.html). Text doesn't reflect that new sowing windows and maturity requirements are now on by default. Fix that, and also add instructions for using hemisphere-wide sowing windows instead.

### LINKED ISSUE #3073: Testmods with nofireemis have it in the wrong order
### Brief summary of bug

The testmods that include nofireemis, have nofireemis LAST, and it should be first in order to get the precedence order right.

This is the reason for most of the issues here:

https://github.com/ESCOMP/CTSM/pull/2844#issuecomment-2811044277

### General bug information

**CTSM version you are using:** ctsm5.3.030-73-gb4298e20e (applies much earlier though)

**Does this bug cause significantly incorrect results in the model's science?** No

**Configurations affected:** tests that need to have fire_emis turned off

### PR DESCRIPTION
**Pull request by ekluzek**
### Description of changes

Various updates for testing and other problems identified in the cesm3_0_beta04 tag. So fixes and cleanup for usability.
Including the following:

- Fix SHR_ASSERT so single-point matrix test passes
- ne3np4 to namelist_defaults_ctsm.xml and Makefile for PTS mode and add ability for warm starts in PTS mode 
- f19 + f45 16pft fsurdat/landuse files to namelist_defaults_ctsm + Makefile
- Changes in the FORTRAN code to properly abort when fire-emission is asked for it can't be provided. Added unit testing for this.

### Specific notes

Contributors other than yourself, if any: @slevis-lmwg 

CTSM Issues Fixed (include github issue #):
  Fixes #2791
  Fixes #2768
  Fixes #2780
  Fixes #2762
  Fixes #3073
  Some of #2810
CTSM namelist checking for: https://github.com/NGEET/fates/issues/1356
Some work on https://github.com/ESCOMP/CTSM/issues/2643

Are answers expected to change (and if so in what way)? No

Any User Interface Changes (namelist or namelist defaults changes)? Yes

Does this create a need to change or add documentation? Did you do so? No 

Testing performed, if any: regular

PR's Involved:

#2840
#2835
#2834
#2844

### CODE REVIEW

**Comment by samsrabin** [comment]
@ekluzek What's the status of this? Is any of it ready to come in / can any of it be split out into different PRs?

**Comment by ekluzek** [comment]
@samsrabin this is my next tag to bring in. We created this branch a long time ago because we thought that the PR's on it would flood b4b-dev too much. But, it turned out the work on it was spread out long enough that that wouldn't have been an issue. At least for this second go around. With b4b-dev we have a clock that ensures changes come in regularly, but we didn't have a process to that here. So it didn't happen. I think that means that in the future we should either slide these things as individual tags to master, or bring them to b4b-dev. Or think about the process that should be in place for a temporary branch like this....

It was brought into this branch as separate PR's similarly to b4b-dev and the first go around with this work. So you can view the contributions singly with those PR's already.

With your ask -- I thought about rebasing the PR's that went into this onto b4b-dev. But at this point I think that would only serve to create more work and slow the work down. So I won't do that now, but this can serve as something to think about as lessons for the future. I'll add this to the Thursday discussion.

**Comment by ekluzek** [comment]
@slevis-lmwg and I just talked to @jtruesdal about this one. We decided we'll bring this in as is. So I'll finish it off. And re-target my other branch that was going to come in for master or b4b-dev.

I had in my mind that my last bit of changes needed to come in with this tag. And I haven't been able to get back to that little bit. I still want/need to finish that off. But, in the spirit of bringing things in when they are ready this part can just come in as is. This is in the spirit of Continuous Integration (CI) from a coding perspective. As well as the freedom we have to bring smaller tags in because of b4b-dev and our having stable versions with minor version updates for scientists. So more small tags is something that we can move toward. There's still some balance because we often have a backlog of tags to go to master. But, this has been outstanding for a long time and we should have brought it in a long time ago.

**Comment by ekluzek** [comment]
I updated to ctsm5.3.057 and testing for aux_clm is passing as expected on Derecho and Izumi.

File: src/biogeochem/CNFireBaseMod.F90 [hunk 1]
```diff
@@ -392,9 +416,11 @@ subroutine FireReadNML( this, NLFilename )
              read(unitn, nml=lifire_inparm, iostat=ierr)
              if (ierr /= 0) then
                 call endrun(msg="ERROR reading "//nmlname//"namelist"//errmsg(sourcefile, __LINE__))
+                return
              end if
           else
              call endrun(msg="ERROR could NOT find "//nmlname//"namelist"//errmsg(sourcefile, __LINE__))
+             return
```
Comment (slevis-lmwg) [inline src/biogeochem/CNFireBaseMod.F90:423]: Do "return" statements serve a purpose after an endrun? (Here and elsewhere.)

File: src/unit_test_stubs/main/ncdio_pio_fake.F90.in [hunk 2]
```diff
@@ -340,6 +341,25 @@ contains
...
+    class(file_desc_t),intent(in) :: ncid      ! PIO file handle
+    character(len=*) , intent(in) :: dimname   ! Dimension name
+    integer, intent(in)           :: value     ! Expected dimension size
+
+    character(len=*), intent(in), optional :: msg  ! Optional additional message printed upon error
+    !
+    ! !LOCAL VARIABLES:
+    !-----------------------------------------------------------------------
+
+    ! Does nothing assumes the dim size is as expected
```
Comment (slevis-lmwg) [inline src/unit_test_stubs/main/ncdio_pio_fake.F90.in:360]: So this subr. is a placeholder for now? I confirmed that a call to it has not been added.

File: src/unit_test_stubs/share_esmf/CMakeLists.txt [hunk 3]
```diff
@@ -1,5 +1,7 @@
 list(APPEND clm_sources
   ExcessIceStreamType.F90
+  FireDataBaseType.F90
+  laiStreamMod.F90
```
Comment (slevis-lmwg) [inline src/unit_test_stubs/share_esmf/CMakeLists.txt:4]: Is laiStreamMod added here as a placeholder for future development or am I misunderstanding?

**Review by slevis-lmwg — APPROVED** [review]
I'm reviewing the part of this that comes in with #2844 since the other PRs were mine.

The title of #2844 predisposed me to expect a short PR with few code changes. This turns out to be a major (in my perception) refactor, plus an addition of unit tests. Thank you for all this @ekluzek!

I included a few questions. Nothing stood out as problematic to me.

File: src/biogeochem/CNFireBaseMod.F90 (hunk 1 above)
Comment (ekluzek) [reply src/biogeochem/CNFireBaseMod.F90:423]: Yes, good question. These are important for the unit testing. In the PFUnit testing the code doesn't terminate after a n endrun call, so it's best to add a return so nothing else happens afterwards.

What I probably *should* DO though, is a add a comment about that. This is done in other places for unit testing, but we might as well explain it each time.

File: src/unit_test_stubs/main/ncdio_pio_fake.F90.in (hunk 2 above)
Comment (ekluzek) [reply src/unit_test_stubs/main/ncdio_pio_fake.F90.in:360]: This is a stub used for the unit testing. The stub module already existed, I just needed to add a new interface that wasn't previously used.

What we are doing in the PFUnit testing is allowing ncdio_pio subroutine calls to be made -- but they do nothing. This removes the problems in unit-testing that filesystem read/writes add. Which is good to do to have very fast solid and robust testing.

File: src/unit_test_stubs/share_esmf/CMakeLists.txt (hunk 3 above)
Comment (ekluzek) [reply src/unit_test_stubs/share_esmf/CMakeLists.txt:4]: The files in the CMakeLists are files that need to be compiled for the unit-tests to work. They require an explicit list, because we can't compile ALL the code for unit-testing.

In this case the list is under the "unit_test_stubs" directory which are the files that have module interfaces of code that's stubbed out in order to fulfill interfaces for compiling/linking, but aren't implemented. So if they are called they don't do anything. Sometimes we add explicit endrun calls in them so that if they are called it'll die -- as actually doing a call of them for unit-testing is a mistake.

**Comment by ekluzek** [comment]
I ran some extra tests to make sure the issues were resolved. These PASS:

PEA.1x1_smallvilleIA.IHistClm50BgcCropQianRs.derecho_gnu.clm-smallville_dynurban_monthly			
PEA_D.1x1_smallvilleIA.IHistClm50BgcCropQianRs.derecho_gnu.clm-smallville_dynurban_monthly			
SEQ_D_PS.f09_f09_mt232.I1850Clm50Sp.derecho_intel.clm-default--clm-nofireemis			
SEQ_PS.f09_f09_mt232.I1850Clm50Sp.derecho_intel.clm-default--clm-nofireemis			
SMS_D_Ln9.f19_f19_mg17.FWma2000climo.derecho_intel.cam-outfrq9s_waccm_ma_mam4			
SMS_D_Ln9_P1280x1.ne0CONUSne30x8_ne0CONUSne30x8_mt12.FCHIST.derecho_intel.cam-outfrq9s			
SMS_D_Ln9_P1280x1.ne0CONUSne30x8_ne0CONUSne30x8_mt12.FCnudged.derecho_intel.cam-outfrq9s			
SMS_Ld12_Mmpi-serial.1x1_vancouverCAN.I1PtClm60SpRs.derecho_gnu.clm-output_sp_highfreq			
SMS_Ln9.ne3pg3_ne3pg3_mg37.I2000Clm50Sp.derecho_gnu.clm-clm50cam6LndTuningMode--clm-nofireemis			
SMS_Ly1_Mmpi-serial.1x1_brazil.IHistClm60BgcQianRs.derecho_intel.clm-output_bgc_highfreq			
SSPMATRIXCN_Ly5_Mmpi-serial.1x1_numaIA.I2000Clm60BgcCropQianRs.derecho_intel.clm-ciso_monthly

These fail and need to be figured out:

ERS_Ld3.f19_f19_mg17.FXHIST.derecho_intel.cam-waccmx_weimer	(SETUP)		
SMS_D_Ln9.f19_f19_mg17.FXHIST.derecho_intel.cam-outfrq9s_amie	(SETUP)		
SMS_D_Ln9_P1280x1.ne0ARCTICne30x4_ne0ARCTICne30x4_mt12.FHIST.derecho_intel.cam-outfrq9s	(NLCOMP RUN)

**Comment by ekluzek** [comment]
It looks like the problem with the f19 tests are the unusual RUN_STARTDATE of: 2005-12-31 and 2003-10-28. It fails with the problem we have been seeing where use_init_interp isn't coordinated correctly with the IC files.

The ARCTIC grid fails in an ESMF regrid and I think it's likely due to too few processors. The default for CAM is 91 nodes, which is 11k tasks, so 10X larger than the tasks asked for. So I think it will likely work with more processors.

**Comment by ekluzek** [comment]
Since, I had some variety of tests failing, I sent the ctsm_sci test list and these tests failed as they need the --clm-nofireemis added as an extra testmod to them.

SMS_Ld5.f09_g17.IHistClm50SpCru.derecho_intel.clm-default       (SETUP)
SMS_Ld5.f19_g17.IHistClm50SpCru.derecho_intel.clm-default       (SETUP)
SMS_Lm12.f09_f09_mg17.I1850Clm60Sp.derecho_intel.clm-ExcessIceStartup_output_sp_exice   (SETUP)
SMS_Lm12.f09_t232.I1850Clm60SpCrujra.derecho_intel.clm-ExcessIceStartup_output_sp_exice (SETUP)

**Comment by ekluzek** [comment]
In the meeting this morning, I was encouraged to make the tag as is. And file issues for anything outstanding. I have created baselines that just need to be renamed. And I'll start the ChangeLog.

**Comment by ekluzek** [comment]
Some tests have different answers because the fieldlists change because fire-emission is off.

ERP_D_Ld3_PS.f09_g17.I2000Clm50Sp.derecho_intel.clm-prescribed		
ERS_D_Ld10.f10_f10_mg37.IHistClm50Sp.derecho_intel.clm-collapse_pfts_78_to_16_decStart_f10		
SMS_C2_D_Lh12.f10_f10_mg37.I2000Clm50Sp.derecho_intel.clm-pauseResume		
SMS_D_Ln9_P128x3.f19_g17.IHistClm50Sp.derecho_intel.clm-waccmx_offline	
ERI_D_Ld9_P48x1.f10_f10_mg37.I2000Clm50Sp.izumi_nag.clm-SNICARFRC		
ERP_D_Ld5_P48x1.f10_f10_mg37.I2000Clm50Sp.izumi_nag.clm-o3lombardozzi2015		
ERS_D.f10_f10_mg37.I1850Clm60Sp.izumi_nag.clm-ExcessIceStreams	

Also the ctsm_sci comparison to ctsm5.3.051 changes answers for VOC's because the MEGAN change came in just after that.
